pkg/controllers/node/tainting/controller.go

```diff
@@ -0,0 +1,139 @@
+	conditionToTaintsMap = map[string]v1.Taint{
+		v1beta1.ConditionTypeDrifted: v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+		v1beta1.ConditionTypeExpired: v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+		v1beta1.ConditionTypeEmpty:   v1beta1.DisruptionCandidatePreferNoScheduleTaint,
+	}
+
+	// managedTaints is the string representation of all taints managed by the tainting controller
+	managedTaints = sets.New(lo.Uniq(lo.MapToSlice(conditionToTaintsMap, func(_ string, taint v1.Taint) string {
+		return taint.ToString()
+	}))...)
+)
+
+func init() {
+	// Assert that no managed taint has a key-value collision. The tainting controller relies on this invariant, change with caution.
+	taintsPerKeyEffect := lo.GroupBy(lo.UniqBy(lo.Values(conditionToTaintsMap), func(taint v1.Taint) string {
+		return taint.ToString()
+	}), func(taint v1.Taint) string {
+		// Group all taints with the same key + effect together
+		taint.Value = ""
+		return taint.ToString()
+	})
+	for _, taints := range taintsPerKeyEffect {
+		if len(taints) != 1 {
+			golog.Fatalf("taints managed by the tainting controller must only have one value per key-effect pair")
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient client.Client
+}
+
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *v1beta1.NodeClaim) (reconcile.Result, error) {
+	node := &v1.Node{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: nodeClaim.Status.NodeName}, node); err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("getting node, %w", err)
+	}
+	stored := node.DeepCopy()
+	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("node", node.Name))
+
+	taintsForConditions := lo.SliceToMap(lo.FilterMap(nodeClaim.GetConditions(), func(cond status.Condition, _ int) (v1.Taint, bool) {
+		if taint, ok := conditionToTaintsMap[cond.Type]; ok {
+			return taint, cond.IsTrue()
+		}
+		return v1.Taint{}, false
+	}), func(taint v1.Taint) (string, v1.Taint) {
+		return taint.ToString(), taint
+	})
+	currentTaints := sets.New(lo.Map(node.Spec.Taints, func(taint v1.Taint, _ int) string {
+		return taint.ToString()
+	})...)
+	toRemove := currentTaints.Intersection(managedTaints.Difference(sets.New(lo.Keys(taintsForConditions)...)))
+	toAdd := sets.New(lo.Keys(taintsForConditions)...)
+	toAdd = toAdd.Difference(currentTaints.Intersection(toAdd))
+
+	if toAdd.Len() == 0 && toRemove.Len() == 0 {
+		return reconcile.Result{}, nil
+	}
+
+	node.Spec.Taints = lo.Reject(node.Spec.Taints, func(taint v1.Taint, _ int) bool {
+		return toRemove.Has(taint.ToString())
+	})
+	node.Spec.Taints = append(node.Spec.Taints, lo.Values(lo.PickByKeys(taintsForConditions, toAdd.UnsortedList()))...)
+
+	if len(toAdd) != 0 {
```

Comment:
Just high-level: This strikes me as complicated and a bit hard to reason about. Are there helpers that we can make here for adding the taint and removing the taint to make this easier for someone new coming in to read?

Reply:
I'm on the same page. Could consider adding a utils taint package that gets utilized in the other places where we're also orchestrating tainting on the node